File: java/src/main/java/com/zhangchi/dsa/ReverseList.java
package com.zhangchi.dsa;

import java.util.ArrayList;
import java.util.Stack;

/**
 * Reverse list
 * @author zhangchi
 * @date 2018/1/15
 */
public class ReverseList {
    public static ListNode AddNode(ListNode head, int val){
        ListNode node = new ListNode(val);
        if (head == null)
            head = node;
        else
        {
            ListNode p = head;
            while (p.next != null)
                p = p.next;

            p.next = node;
        }
        return head;
    }

    // Reverse list
    public static ListNode Reverse(ListNode head){
        if(head == null)
            return head;

        ListNode prev = null;
        ListNode cur = head;
        ListNode temp;

        while(cur != null){
            temp = cur.next;
            cur.next = prev;
            prev = cur;
            cur = temp;
        }

        return prev;
    }

    public static void printList(ListNode head){
        while (head != null)
        {
            System.out.print(head.val + " ");
            head = head.next;
        }
        System.out.println();
    }

    public static <T> void printArray(T[] arr){
        for(T e : arr)
            System.out.print(e + " ");
        System.out.println();
    }


    // 从尾到头打印链表（nowcoder,3）
    public static ArrayList<Integer> printListFromTailToHead(ListNode head){
        if (head == null)
            return  null;
        ArrayList<Integer> arrayList = new ArrayList<Integer>();
        Stack<Integer> stack = new Stack<Integer>();

        ListNode p = head;
        while(p != null)
        {
            stack.push(p.val);
            p = p.next;
        }

        while (!stack.empty())
            arrayList.add(stack.pop());

        return arrayList;
    }

    public static void main(String[] args){
        ListNode head = new ListNode(1);
        ReverseList.AddNode(head, 2);
        ReverseList.AddNode(head, 3);
        ReverseList.AddNode(head, 4);
        printArray(printListFromTailToHead(head).toArray());
        printList(head);
        ListNode newHead = ReverseList.Reverse(head);
        printList(newHead);

        return;
    }
}

class ListNode{
    int val;
    ListNode next;

    ListNode(int val){
        this.val = val;
    }
}

File: cpp/ReverseList.cpp
#include <iostream>

using namespace std;

struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* ReverseList(ListNode* pHead) {
		if (pHead == NULL)
			return NULL;

		ListNode * prev = NULL;
		ListNode * cur = pHead;
		ListNode * temp = NULL;

		while (cur != NULL)
		{
			temp = cur->next;
			cur->next = prev;
			prev = cur;
			cur = temp;
		}

		return prev;
	}
};

class Solution_Recursion {
public:
	ListNode* ReverseList(ListNode* pHead) {
		if (!pHead || !pHead->next) return pHead;
		ListNode* last = ReverseList(pHead->next);
		pHead->next->next = pHead;
		pHead->next = NULL;

		return last;
	}	
};

int main()
{
	ListNode* cur = nullptr;
	ListNode* head = cur = new ListNode(1);
	cur = cur->next = new ListNode(2);
	cur = cur->next = new ListNode(3);
	printList(head);

	Solution_Recursion SR;
	ListNode* reverseHead = SR.ReverseList(head);
	printList(reverseHead);

	return 0;
}